db_setup/app/__init__.py

```diff
@@ -16,63 +17,137 @@
-            template = "An SQLAlchemy exception of type {0} occurred. Arguments:\n{1!r}"
-            print(resource)
-            break
-        except Exception as e:
-            db.session.rollback()
-            print('exception', e)
-            template = "An exception of type {0} occurred. Arguments:\n{1!r}"
-            print(resource)
-            break
+        category = get_category(resource, category_dict) # Note: modifies the category_dict in place (bad?)
+        langs = get_languages(resource, language_dict) # Note: modifies the language_dict in place (bad?)
+        existing_resource = existing_resources.get(resource['url'])
 
+        if existing_resource:
+            update_resource(resource, existing_resource, langs, category)
+        else:
+            create_resource(resource, langs, category)
 
+def remove_duplicates(data):
+    unique_resources = []
+    resource_dict = {}
+    for resource in data:
+        if not resource_dict.get(resource['url']):
+            resource_dict[resource['url']] = True
+            unique_resources.append(resource)
+    return unique_resources
+
+def get_category(resource, category_dict):
+    category = resource.get('category')
+
+    if category not in category_dict:
+        category_dict[category] = Category(name=category)
+
+    return category_dict[category]
+
+def get_languages(resource, language_dict):
+    langs = []
+
+    # Loop through languages and create a new Language
+    # object for any that don't exist in the DB
+    for language in resource.get('languages') or []:
+        if language not in language_dict:
+            language_dict[language] = Language(name=language)
+
+        # Add each Language object associated with this resource
+        # to the list we'll return
+        langs.append(language_dict[language])
+
+    return langs
+
+def create_resource(resource, langs, category):
+    try:
+        new_resource = Resource(
+            name=resource['name'],
+            url=resource['url'],
+            category=category,
+            languages=langs,
+            paid=resource.get('paid'),
+            notes=resource.get('notes', ''),
+            upvotes=resource.get('upvotes', 0),
+            downvotes=resource.get('downvotes', 0),
+            times_clicked=resource.get('times_clicked', 0))
+
+        db.session.add(new_resource)
+        db.session.commit()
+    except exc.SQLAlchemyError as e:
+        db.session.rollback()
+        print('Flask SQLAlchemy Exception:', e)
+        template = "An SQLAlchemy exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+    except Exception as e:
+        db.session.rollback()
+        print('exception', e)
+        template = "An exception of type {0} occurred. Arguments:\n{1!r}"
+        print(resource)
+
+def update_resource(resource, existing_resource, langs, category):
+    # Return without updating if the resource is already up to date
```

Comment:
should we do 
`
resource['languages'] = langs
`
and then change method:
`
match_resource(resource, existing_resource)
`

Reply:
Ah, yes good call. This became obvious once I refactored to the point where I'm at. It was less obvious to do it this way with the original code. Now that you point it out, I'm going to implement that